lib/core/src/Cardano/Wallet/Address/Pool.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Copyright: © 2021 IOHK
+-- License: Apache-2.0
+--
+-- An address pool caches a collection of addresses.
+-- The purpose of this data structure is to aid in BIP-44 style
+-- address discovery with an address gap.
+module Cardano.Wallet.Address.Pool
+    ( Pool
+    , generator
+    , addresses
+    , gap
+    , lookup
+    , size
+    , next
+    , new
+    , load
+    , update
+    , clear
+
+    -- * Internal
+    , loadUnsafe
+    , prop_sequence
+    , prop_gap
+    , prop_fresh
+    , prop_generator
+    , prop_consistent
+    )
+  where
+
+import Prelude hiding
+    ( last, lookup )
+
+import Cardano.Wallet.Primitive.Types.Address
+    ( AddressState (..) )
+import Control.DeepSeq
+    ( NFData )
+import Data.Map.Strict
+    ( Map )
+import Data.Ord
+    ( Down (..) )
+import Fmt
+    ( Buildable (..) )
+import GHC.Generics
+    ( Generic )
+
+{- HLINT ignore "Avoid restricted qualification" -}
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+    Address Pool, abstract data type
+-------------------------------------------------------------------------------}
+-- | An address pool caches a collection of addresses (type @addr@)
+-- which are derived from a numeric index (type @ix@).
+data Pool addr ix = Pool
+    { generator :: ix -> addr
+    -- ^ Each address is obtained from a numeric index.
+    -- The purpose of the 'Pool' data structure is to (partially)
+    -- cache this mapping,
+    -- because it is expensive to compute (hashing)
+    -- and its inverse is practically impossible to compute.
+    , gap :: Int
+    -- ^ The pool gap determines how 'Used' and 'Unused'
+    -- have to be distributed.
+    -- See 'prop_gap' and 'prop_fresh'.
+    , addresses :: Map addr (ix, AddressState)
+    -- ^ Partial, cached inverse of the 'generator'.
+    -- This map contains all cached addresses @addr@,
+    -- their corresponding indices @ix@,
+    -- and whether they are 'Used' or 'Unused'.
+    -- See 'prop_sequence'.
+    } deriving (Generic)
+
+instance (NFData addr, NFData ix) => NFData (Pool addr ix)
+
+-- | Internal invariant:
+-- The indices of the addresses in a pool form a finite
+-- sequence beginning with 'fromEnum'@ 0@.
+prop_sequence :: (Ord ix, Enum ix) => Pool addr ix -> Bool
+prop_sequence Pool{addresses} =
+    indices `List.isPrefixOf` [toEnum 0..]
+  where
+    indices = List.sort $ map fst $ Map.elems addresses
+
+-- | Internal invariant:
+-- If we order the 'addresses' by their indices,
+-- then there are always /less than/ 'gap' many 'Unused'
```

Comment:
we should have EXACTLY `gap` number of Unused addresses "at the end of indices list", right?

Reply:
Yes, but `prop_gap` only makes a statement about addresses "between two consecutive 'Used' addresses" (next line), and that is never "at the end of indices list". So, `prop_gap` holds. 🤓 The end of the address list is captured by `prop_fresh`.